index.html

```
@@ -140,67 +166,250 @@ <h2>
+          <a data-cite="html#responsible-document">responsible document</a>.
+          </li>
+          <li>If <var>contents</var> is omitted, set the badge associated with
+          <var>document</var> to <a>flag</a>.
           </li>
-          <li>If <var>contents</var> is 0, set the application's badge to
-          <a>nothing</a>.
+          <li>Else, if <var>contents</var> is 0, set the badge associated with
+          <var>document</var> to <a>nothing</a>.
           </li>
-          <li>Else, set the application's badge to <var>contents</var>.
+          <li>Else, set the badge associated with <var>document</var> to <var>
+            contents</var>.
+          </li>
+          <li>Return a <a data-cite="webidl#a-promise-resolved-with">resolved
+          promise</a>.
+          </li>
+        </ol>
+      </section>
+      <section>
+        <h3>
+          <dfn data-dfn-for="Navigator">clearClientBadge()</dfn> method
+        </h3>
+        <p>
+          When the <code>clearClientBadge()</code> method is called, set the
+          the badge associated with the <a data-cite=
+          "html#current-settings-object">current settings object</a>'s
+          <a data-cite="html#responsible-document">responsible document</a> to
+          <a>nothing</a>, and return a <a data-cite=
+          "webidl#a-promise-resolved-with">resolved promise</a>.
+        </p>
+      </section>
+      <section>
+        <h3>
+          <dfn data-dfn-for="NavigatorBadge">setAppBadge()</dfn> method
+        </h3>
+        <p>
+          When the <code>setAppBadge()</code> method is called with argument
+          <var>contents</var>, run the following steps:
+        </p>
+        <ol>
+          <li>Let <var>result</var> be the result of <a>determining the
+          matching applications</a>. If <var>result</var> was a
+          {{DOMException}}, return a <a data-cite=
+          "webidl#a-promise-rejected-with">rejected promise</a> with reason
+          <var>result</var>.
```

Hmm, should we reject or fail silently? This could kind of leak installed state to the site (though it's already kind of leaked to them through `appinstalled` and `display-mode`). I don't know if we want to make the badging API a proxy `isInstalled()` function..

---

Having read the `determining matching applications` I think this comment is wrong. Rejection only occurs in Workers that aren't the service worker, correct?